### PR TITLE
[Source Gen] Revise prop for enabling generation and add new condition

### PR DESCRIPTION
**Pull request by satvu**
### Issue describing the changes in this PR

Enable generation only if worker-indexing is enabled.

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

**Comment by satvu** [comment]
Still need to validate locally - but will be done EOD

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets [hunk 1]
```diff
@@ -34,7 +34,7 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
         <MSBuildFunctionsTargetsPath>$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed.Functions\</MSBuildFunctionsTargetsPath>
       
         <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnableWorkerIndexing) == ''">false</FunctionsEnableWorkerIndexing>
-        <FunctionsMetadataSourceGen_Enabled Condition="$(FunctionsMetadataSourceGen_Enabled) == ''">true</FunctionsMetadataSourceGen_Enabled>
+        <FunctionsEnableMetadataSourceGen Condition="$(FunctionsEnableWorkerIndexing) == 'true' And $(FunctionsEnableMetadataSourceGen) == ''">true</FunctionsEnableMetadataSourceGen>
```
[the remark this answers is not in the document]
Comment (satvu) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:37]: Will add!

**Comment by satvu** [comment]
Validation Scenarios (w/ MSBuild Log Viewer)

- [x] `<FunctionsEnablePlaceholder>true<FunctionsEnablePlaceholder>` enables both `FunctionsEnableWorkerIndexing` and `FunctionsEnableMetadataSourceGen` (see: https://github.com/Azure/azure-functions-dotnet-worker/pull/1258, placeholder will always use WorkerIndexing)
- [x] All `FunctionsEnable` prefixed props are false when not explicitly set
- [x] `FunctionsEnableMetadatSourceGen` is only set to true if no value has been set AND `FunctionsEnableWorkerIndexing` is true

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets [hunk 2]
```diff
@@ -33,9 +33,10 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
         <_FunctionsExtensionsFullPublish Condition="$(_FunctionsExtensionsFullPublish) == ''">!$(NoBuild)</_FunctionsExtensionsFullPublish>
         <MSBuildFunctionsTargetsPath>$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed.Functions\</MSBuildFunctionsTargetsPath>
       
-        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnableWorkerIndexing) == ''">false</FunctionsEnableWorkerIndexing>
-        <FunctionsMetadataSourceGen_Enabled Condition="$(FunctionsMetadataSourceGen_Enabled) == ''">true</FunctionsMetadataSourceGen_Enabled>
         <FunctionsEnablePlaceholder Condition="$(FunctionsEnablePlaceholder) == ''">false</FunctionsEnablePlaceholder>
+        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnablePlaceholder) == 'true'">true</FunctionsEnableWorkerIndexing>
+        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnableWorkerIndexing) == '' And !$(FunctionsEnablePlaceholder)">false</FunctionsEnableWorkerIndexing>
+        <FunctionsEnableMetadataSourceGen Condition="$(FunctionsEnableWorkerIndexing) == 'true' And $(FunctionsEnableMetadataSourceGen) == ''">true</FunctionsEnableMetadataSourceGen>
```
Comment (satvu) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:39]: Another question - do we want to forcibly set `FunctionsEnableMetadataSourceGen` to false if `FunctionsEnableWorkerIndexing` is not enabled?

Comment (kshyju) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:39]: > Another question - do we want to forcibly set `FunctionsEnableMetadataSourceGen` to false if `FunctionsEnableWorkerIndexing` is not enabled?

Do we need both the settings? Why not just "FunctionsEnableWorkerIndexing"?

Comment (satvu) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:39]: https://github.com/Azure/azure-functions-dotnet-worker/issues/1155 - `FunctionsEnableMetadataSourceGen` was introduced so people could turn off generation in cases of compilation errors. 

There is a path for worker-indexing without source-generation (reads `functions.json` after worker initialization) which would be taken if worker indexing was true and source generation was false. Not sure which scenarios someone would have to use this.

If we want worker indexing to _only_ use source-generation, we can cut down to just one prop.

Comment (kshyju) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:39]: got it. May be we can re-evaluate this in few months again?

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets [hunk 3]
```diff
@@ -33,9 +33,10 @@ WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and
         <_FunctionsExtensionsFullPublish Condition="$(_FunctionsExtensionsFullPublish) == ''">!$(NoBuild)</_FunctionsExtensionsFullPublish>
         <MSBuildFunctionsTargetsPath>$(MSBuildExtensionsPath)\Microsoft\VisualStudio\Managed.Functions\</MSBuildFunctionsTargetsPath>
       
-        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnableWorkerIndexing) == ''">false</FunctionsEnableWorkerIndexing>
-        <FunctionsMetadataSourceGen_Enabled Condition="$(FunctionsMetadataSourceGen_Enabled) == ''">true</FunctionsMetadataSourceGen_Enabled>
         <FunctionsEnablePlaceholder Condition="$(FunctionsEnablePlaceholder) == ''">false</FunctionsEnablePlaceholder>
+        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnablePlaceholder) == 'true'">true</FunctionsEnableWorkerIndexing>
+        <FunctionsEnableWorkerIndexing Condition="$(FunctionsEnableWorkerIndexing) == '' And !$(FunctionsEnablePlaceholder)">false</FunctionsEnableWorkerIndexing>
```
Comment (brettsam) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:38]: in one line we check `$(FunctionsEnablePlaceholder) == 'true'` and in the next we do `!$(FunctionsEnablePlaceholder)` -- do both work? if so, using the `!` is more readable, i'd think...

Comment (satvu) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:38]: Both work - are you suggesting changing `$(FunctionsEnablePlaceholder) == 'true'` to just `$(FunctionsEnablePlaceholder)`?

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets (hunk 2 above)
Comment (satvu) [reply sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:39]: Will add #2 - I think 3 can come in a follow up.

**Review by surgupta-msft — APPROVED** [review]

**Review by brettsam — APPROVED** [review]

**Pull request merged**